lib/iris/experimental/regrid.py

```diff
@@ -473,140 +473,251 @@ def _regrid_area_weighted_array(
+                    if isinstance(x_indices, tuple) and isinstance(
+                        y_indices, tuple
+                    ):
+                        raise RuntimeError(
+                            "Cannot handle split bounds " "in both x and y."
+                        )
+                    weights = area_func(y_bounds, x_bounds)
+                weights_i.append(weights)
+            cached_weights.append(weights_i)
+        return (
+            tuple(cached_x_indices),
+            tuple(cached_y_indices),
+            max_x_indices,
+            max_y_indices,
+            tuple(cached_weights),
+        )
+
+    weights_info = _calculate_regrid_area_weighted_weights(
+        src_x_bounds,
+        src_y_bounds,
+        grid_x_bounds,
+        grid_y_bounds,
+        grid_x_decreasing,
+        grid_y_decreasing,
+        area_func,
+        circular,
+    )
+    (
+        cached_x_indices,
+        cached_y_indices,
+        max_x_indices,
+        max_y_indices,
+        cached_weights,
+    ) = weights_info
+    # Delete variables that are not needed and would not be available
+    # if _calculate_regrid_area_weighted_weights was refactored further
+    del src_x_bounds, src_y_bounds, grid_x_bounds, grid_y_bounds
+    del grid_x_decreasing, grid_y_decreasing
+    del area_func, circular
+
+    # Ensure we have x_dim and y_dim.
+    x_dim_orig = copy.copy(x_dim)
+    y_dim_orig = copy.copy(y_dim)
+    if y_dim is None:
+        src_data = np.expand_dims(src_data, axis=src_data.ndim)
+        y_dim = src_data.ndim - 1
+    if x_dim is None:
+        src_data = np.expand_dims(src_data, axis=src_data.ndim)
+        x_dim = src_data.ndim - 1
+    # Move y_dim and x_dim to last dimensions
+    src_data = np.moveaxis(src_data, x_dim, -1)
+    if x_dim < y_dim:
+        src_data = np.moveaxis(src_data, y_dim - 1, -2)
+    elif x_dim > y_dim:
+        src_data = np.moveaxis(src_data, y_dim, -2)
+    x_dim = src_data.ndim - 1
+    y_dim = src_data.ndim - 2
+
+    # Create empty "pre-averaging" data array that will enable the
+    # src_data data coresponding to a given target grid point,
+    # to be stacked per point.
+    # Note that dtype is not preserved and that the array mask
+    # allows for regions that do not overlap.
+    new_shape = list(src_data.shape)
+    new_shape[x_dim] = len(cached_x_indices)
+    new_shape[y_dim] = len(cached_y_indices)
+    num_target_pts = len(cached_y_indices) * len(cached_x_indices)
+    src_areas_shape = list(src_data.shape)
+    src_areas_shape[y_dim] = max_y_indices
+    src_areas_shape[x_dim] = max_x_indices
+    src_areas_shape += [num_target_pts]
     # Use input cube dtype or convert values to the smallest possible float
     # dtype when necessary.
     dtype = np.promote_types(src_data.dtype, np.float16)
+    # Create empty arrays to hold src_data per target point, and weights
+    src_area_datas = np.zeros(src_areas_shape, dtype=np.float64)
```

Comment:
Should `dtype` above be used rather than `np.float64`?

Reply:
I think as it standard the code is using the numpy default value of np.float64, so I've used this here.  I don't think we should revisit the dtype handling in a follow up PR as I think it would be beneficial to improve it.